CMakeLists.txt

```diff
@@ -119,91 +128,151 @@ set(BOOST_MODULES_TO_FIND filesystem system)
-		                  else()
-		                      message(WARNING "Boost compilation on Windows requires the use of a clang-cl compiler running inside a Visual Studio Developer CLI environment with a Windows C++ SDK available.")
-		                      set(BOOST_TOOLSET "msvc")
-		                  endif()
-		              else()
-		                  set(BOOST_TOOLSET "msvc")
-		              endif()
-		          elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "Linux")
-		              if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-		                  set(BOOST_TOOLSET "clang")
-		              elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-		                  set(BOOST_TOOLSET "gcc")
-		              endif()
-		          endif()
-
-		          # Determine b2 executable name depending on host system
-		          if(CMAKE_HOST_SYSTEM_NAME STREQUAL "Windows")
-		              set(b2_EXECUTABLE_NAME "b2.exe")
-		          else()
-		              set(b2_EXECUTABLE_NAME "b2")
-		          endif()
-
-		          # Find b2
-		          find_program(b2 NAME ${b2_EXECUTABLE_NAME} PATHS ${BOOST_SOURCEDIR} NO_CACHE NO_DEFAULT_PATH)
-
-		          # Compile b2 if it isn't found
-		          if(b2 STREQUAL "b2-NOTFOUND")
-		              message(STATUS "The b2 buildsystem executable hasn't been found in your system. Compiling b2...")
-		              b2_build()
-
-		              # Find b2 again after compilation
-		              find_program(b2 NAME ${b2_EXECUTABLE_NAME} PATHS ${BOOST_SOURCEDIR} NO_CACHE NO_DEFAULT_PATH REQUIRED)
-		          endif()
-
-		          # Compile Boost
-		          message(STATUS "Compiling Boost...")
-		          boost_compile()
-
-		          # Find Boost again
-		          find_package(Boost COMPONENTS ${BOOST_MODULES_TO_FIND} PATHS ${BOOST_INSTALLDIR} NO_DEFAULT_PATH REQUIRED)
-		      else()
-		          message(STATUS "Custom Boost distribution found in ${BOOST_INSTALLDIR}")
-		      endif()
-		  endif()
-	else()
-		  # Try to find Boost on the system and CMake's default paths
-		  set(Boost_USE_STATIC_LIBS ON)
-		  find_package(Boost COMPONENTS ${BOOST_MODULES_TO_FIND} REQUIRED)
-	endif()
+    # Windows (Specifically MSVC, but it technically applies to all builds regardless of compiler / OS...) doesn't handle
+    # symbol redefinitions gracefully.
+    #
+    # In the case of Boost, BoostConfig.cmake doesn't use CMake's generator expressions when defining it's link libraries,
+    # instead adding the found definitions to the project's link dependencies directly. With no way to perform library detection only.
+    # Worse, get_boost() and boost_compile() runs during configure so we don't have a build configuration yet, and therefore do not know what
+    # build of Boost (Release or Debug) to use at that point. This means that calling find_package(Boost ...) more than once can result
+    # in a project that has both the Debug and Release libraries added as dependencies. (Which will cause a link failure under MSVC.
```

Comment:
How can boost try to use both a Debug and Release build? It will by default use a build matching the current target build type

Reply:
When cmake is initially called with the preset option, the build configuration isn't set by VS2022. As such everything winds up defaulting to the Debug build as that's the first entry in cmake's default configuration list. This causes Boost to add it's debug dependencies to the build when get_boost() calls find_package().

Later, when cmake is invoked again with the build option, the config option is also given. This sets the build configuration to a specific type. (Whatever the user selected in VS2022's GUI.) As long as that configuration was "Debug" the build will succeed. As get_boost()'s call to find_package() will effectively be a no-op.

If the user has selected a different build configuration however, then the repeated call to find_package() will append a new Boost library and set of include headers to the target. VS2022 can not handle this, and thus we get Vita3K/Vita3K#3092.

Re-running the build, *without changing the build configuration in VS2022's GUI*, however will get rid of the old incorrect dependency and allow the build to succeed, but that's inconvenient to those trying to follow the manual build instructions for Windows. We also cannot remove the incorrect dependency ourselves once it's added, as cmake doesn't provide a reliable option / command to do so. As such our only option is to prevent the second call to find_package() from being made in the first place.

As to my theory as to why this behavior happens and why the work around, well, works: 

In my build of Vita3K's external boost::filesystem we have these files at Vita3K\build\windows-vs2022\external\boost\lib\cmake\boost_filesystem-1.82.0:
boost_filesystem-config.cmake
boost_filesystem-config-version.cmake
libboost_filesystem-variant-vc143-mt-gd-x64-1_82-static.cmake
libboost_filesystem-variant-vc143-mt-x64-1_82-static.cmake

The actual call to add_library() / set_target_properties() is in the variant files. (in both variants at line 75 & 77 for me.) With the config files effectively re-implementing cmake's generator expressions via the various flags: Boost_USE_SINGLE_VARIANT / Boost_USE_RELEASE_LIBS / Boost_USE_DEBUG_LIBS / etc. (I guess this predates cmake's generator expressions? But there is limited use of them....)

There *is* code in there to block calling add_library() / set_target_properties() using the same suggested change you made above, i.e. if(NOT TARGET Boost::filesystem), but it fails here. My guess is that cmake isn't keeping all of the information between cmake invocations because of the changed build configuration type, but cmake *is* reusing the previous list of libraries and includes for each project as a template. As such, if (NOT TARGET ...) fails as cmake needs to reconfigure it based on the new build configuration. Which makes sense. That's the entire point of cmake's generator expressions. Cmake's preset stage is used to generate the default dependency list. Then cmake's build stage, with it's defined build configuration type, is only supposed to append what is specifically needed for that build configuration.